---
title: "Figure 5"
output: 
  html_document:
    df_print: paged
author: "John Favate"
date: "`r Sys.time()`"
---

<style type="text/css">
.main-container {
  max-width: 1500px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r}
# Prevent printing of warnings and such in the HTML
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.align = "center")
```

```{r}
library(tidyverse)
library(ggpubr)
library(patchwork)
library(rtracklayer)
library(ggridges)

all.data <- read_csv("../../data_frames/all_data.csv") %>%
  filter(gene_type == "ECB")

mut.levels <- all.data %>%
  select(line, mutator) %>%
  arrange(mutator, line) %>%
  unique() %>%
  pull(line)

mut.colors <- c(rep("black", 6), rep("firebrick3", 5))

mutations <- read_csv("../../data_frames/table_s8_mutations.csv")%>%
  mutate(clone_id = str_remove(strain, "REL")) %>%
  filter(population != "Ara+6")

my.clones <- c("11330","11392","11333","11342","11345","11364","11336","11348","11339","11367","11370","11389")

my.mutations <- mutations %>%
  filter(clone_id %in% my.clones & time == 50000)
```

Order the genes in all data by mean fc across the lines
```{r}
gene.order <- all.data %>% 
  group_by(target_id) %>% 
  summarise(mean_fc = mean(ds_log2foldchange_rna, na.rm = TRUE)) %>% 
  arrange(mean_fc) %>% 
  pull(target_id)

all.data$target_id <- factor(all.data$target_id, levels = gene.order)

# also keep mutator order
all.data$line <- factor(all.data$line, levels = mut.levels)
```

We marked any gene with any kind of deletion as "deleted", though this actually encompasses genes which are entirely deleted and those that have partial deletions. These can be differentiated by whether they are present in the GFF or not. 
```{r}
gff.locs <- dir("../../gffs", full.names = TRUE)

names(gff.locs) <- str_extract(gff.locs, "Ara[MPR][1-9]") %>% 
  str_replace("M", "-") %>% 
  str_replace("P", "+") %>% 
  str_replace("AraR", "REL0")

gff.list <- lapply(gff.locs, readGFFAsGRanges)
```

Find all the genes that are fCDS, i.e. pseudogenized, i.e. containing partial deletions. 
```{r}
partial.genes.list <- lapply(gff.list[names(gff.list) != "AraR6"], function(x){
  evo.genes <- x[x$type == "fCDS"]$ID %>%
    str_extract("ECB_[0-9]{5}") %>%
    unique()
  
  return(data.frame(target_id = evo.genes))
})

(partial.genes.df <- bind_rows(partial.genes.list, .id = "line") %>% 
    mutate(del_type = "partial"))
```

When I join that data frame to all data, genes that are partials will have a match, but complete deletions will be NA. Technically, all other genes will be NA, even non-deleted genes but that can be fixed easily.
```{r}
dels.df <- all.data %>%
  filter(gene_type == "ECB") %>% 
  select(line, target_id, deleted, ds_log2foldchange_ribo, ds_log2foldchange_rna, ds_padj_rna, ds_padj_ribo, anc_mean_tpm_rna, anc_mean_tpm_ribo) %>%
  left_join(., partial.genes.df, by = c("line", "target_id")) %>%
  replace_na(list(del_type = "complete")) %>% 
  mutate(del_type = ifelse(deleted == TRUE, del_type, "present"))

head(dels.df)
```

The collective set of genes with any deletion.
```{r}
collective.del <- dels.df %>% 
  filter(deleted == TRUE) %>% 
  pull(target_id) %>% 
  unique()
```

One issue is that many genes are fully deleted in some lines and partially in another. This df shows the number of times a gene is either complete, partly, or not deleted across the lines for the collective set of genes that was deleted at least once in any line. 
```{r}
del.type.df <- dels.df %>% 
  filter(target_id %in% collective.del) %>% 
  group_by(del_type, target_id) %>% 
  tally() %>% 
  ungroup() %>% 
  pivot_wider(names_from = del_type, values_from = n) %>% 
  replace_na(list(complete = 0, partial = 0, present = 0)) %>% 
  mutate(total_dels = partial + complete,
         comp_over_partial = ifelse(is.infinite(complete/partial), complete, complete/partial)) %>% 
  arrange(desc(total_dels), desc(comp_over_partial))

del.type.order <- del.type.df %>% 
  pull(target_id)

del.type.df$target_id <- factor(del.type.df$target_id, levels = del.type.order)

head(del.type.df)
```

Sets of genes we're interested in
```{r}
# completed deleted in at least 4 lines
comp.4.or.more <- del.type.df %>% 
  filter(complete >= 4) %>% 
  pull(target_id)

# never deleted
never.del <- all.data %>% 
  pull(target_id) %>% 
  unique() %>% 
  setdiff(., collective.del)
```

### Panel A

Number of genes completely deleted in N lines
```{r}
pa <- del.type.df %>% 
  filter(complete != 0) %>% 
  group_by(complete) %>% 
  tally() %>% 
  ungroup() %>% 
  add_case(complete = 11, n = 0) %>% 
  ggplot(., aes(complete, n))+
  geom_col()+
  scale_x_continuous(breaks = 1:11)+
  theme_classic()+
  theme(text = element_text(size = 13))+
  labs(x = "Number of lines in which a\ngene was completely deleted",
       y = "Number of genes",
       title = "A")

pa
```

### Panel B

```{r fig.width = 4.5, fig.height = 4}
l2 <- list("a_never" = never.del,
           "b_del.4.or.more" = comp.4.or.more)

anctpmdf <- lapply(l2, function(x){
  all.data %>% 
    filter(target_id %in% x) %>%
    select(target_id, anc_mean_tpm_rna, anc_mean_tpm_ribo) %>% 
    unique()
}) %>% 
  bind_rows(.id = "class") %>% 
  pivot_longer(c(anc_mean_tpm_rna, anc_mean_tpm_ribo))

anctpmdf$name <- factor(anctpmdf$name, levels = c("anc_mean_tpm_rna", "anc_mean_tpm_ribo"))

pb <- anctpmdf %>% 
  ggplot(., aes(name, value, fill = class))+
  geom_boxplot()+
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x)),
                breaks = c(10^-2, 10^0, 10^2, 10^4),
                limits = c(10^-3, 10^5))+
  scale_fill_manual(values = c("grey70", "tomato"), 
                    labels = c("Never deleted", paste("Deleted in \u2265 4 lines")),
                    name = NULL)+
  scale_x_discrete(labels = c("RNA-seq", "Ribo-seq"))+
  theme_classic()+
  theme(text = element_text(size = 13),
        legend.position = c(.5, .03),
        legend.direction = "horizontal",
        legend.background = element_blank())+
  labs(x = NULL,
       y = "Ancestral TPM",
       title = "B")+
  stat_compare_means(aes(label = paste("p =", ..p.format..)), label.y.npc = 1)

pb
```

### Panel C

Panel A will be a compound panel with two things, the bottom is a fold change heatmap, above it is the frequency of deletion. The heatmap part:
```{r}
# the color scale labels
scale.labels <- c(paste("\u2264", paste0("-", 3), sep = ""),
                  "",
                  0,
                  "",
                  paste("\u2265", 3, sep = ""))

# create data frame for panel A
adf <- all.data %>% 
  filter(target_id %in% comp.4.or.more)

# make a max fc of 3 in either direction
adf$ds_log2foldchange_rna[adf$ds_log2foldchange_rna > 3] <- 3

adf$ds_log2foldchange_rna[adf$ds_log2foldchange_rna < -3] <- -3

pc.bottom <- adf %>% 
  ggplot(., aes(target_id, line, fill = ds_log2foldchange_rna))+
  geom_raster()+
  scale_fill_gradientn(colours = c("firebrick2", "white", "dodgerblue2"),
                       name = expression(paste(log[2], "(fold-change)", sep = "")),
                       labels = scale.labels,
                       breaks = c(-3, (-3/2), 0, (3/2), 3),
                       guide = guide_colorbar(ticks.colour = "black"),
                       limits = c(-3.1, 3.1))+
  theme(panel.background = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        axis.text.y = element_text(color = mut.colors),
        text = element_text(size = 13),
        legend.position = "right",
        legend.direction = "vertical")+
  labs(x = "Transcripts",
       y = NULL)

pc.bottom
```

The frequency of deletion of each of those genes. **One issue is that the two genes on the right are completely deleted 10 times, and partial in the remaining line, they still have no fold change so it's 11 grey boxes but the bar will only show 10 complete dels for each**
```{r}
adf2 <- del.type.df %>% 
  filter(target_id %in% comp.4.or.more) %>% 
  mutate(target_id = factor(target_id, levels = gene.order))

pc.top <- adf2 %>% 
  ggplot(., aes(target_id, complete-3))+
  geom_col(width = 1)+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.background = element_blank(),
        panel.grid = element_blank(),
        axis.line.y = element_line(color = "black"),
        text = element_text(size = 13))+
  scale_y_continuous(expand = expansion(mult = c(0, .1)), 
                     breaks = seq(1, 7, 3), labels = c(4, 7, 10))+
  labs(x = NULL, 
       y = "Frequency of\ncomplete deletion",
       title = "C")

pc.top
```

The complete panel C
```{r}
pc <- (pc.top / pc.bottom)
```


### Panel D

A panel showing the number of genes afflicted by various deletions
```{r}
afflicted.genes <- my.mutations %>% 
  filter(mutation_category == "large_deletion") %>% 
  pull(gene_list)

# find mean number of genes hit
mean.genes.hit <- sapply(afflicted.genes, function(x){
  str_split(x, ",") %>% 
    lengths()
}) %>% 
  mean()

pd <- sapply(afflicted.genes, function(x){
  str_split(x, ",") %>% 
    lengths()
}) %>% 
  as_tibble() %>% ggplot(., aes(value))+
  geom_histogram(bins = 40)+
  theme_classic()+
  theme(text = element_text(size = 13))+
  labs(x = "Number of genes deleted per large deletion",
       y = "Frequency",
       title = "D")+
  scale_y_continuous(breaks = seq(0, 20, 5), limits = c(0, 20))+
  scale_x_continuous(limits = c(0, 200))+
  geom_vline(aes(xintercept = mean.genes.hit), linetype = 5)+
  annotate("text",label = paste("~", floor(mean.genes.hit), "genes"), x = 50, y = 19)

pd
```

### Final figure

```{r fig.width = 10, fig.height = 8}
top <- pa | pb

bottom <- pc | pd

(ff <- (top / bottom))
```

Save it
```{r}
ggsave(ff, filename = "../../figures/f_fig_5.pdf", width = 10, height = 8, device = cairo_pdf)

ggsave(ff, filename = "../../figures/f_fig_5.png", width = 10, height = 8, dpi = 300)
```

```{r}
sessionInfo()
```

